### PR TITLE
HWKAGENT-125 Use server-level operation commands for deployments

### PR DESCRIPTION
**Pull request by jshaughn**
- use appropriate deployment terminology in the code based on
  our provider conventions.
- add new options to deploy/undeploy
- add support for enable/disable/restart deploymemnt requests

### CODE REVIEW

File: lib/hawkular/operations/operations_api.rb [hunk 1]
```diff
@@ -128,46 +128,113 @@ def invoke_specific_operation(operation_payload, operation_name, &callback)
...
-    # @option hash [String]  :resource_path canonical path of the WildFly server from which to remove the deployment
-    # @option hash [String]  :deployment_name name of deployment to remove
+    # @option hash [String]  :resource_path canonical path of the WildFly server from which to undeploy the deployment
+    # @option hash [String]  :deployment_name name of deployment to undeploy
+    # @option hash [String]  :remove_content whether to remove the deployment content or not (default = true)
+    # @option hash [String]  :server_groups comma-separated list of server groups for the operation (default = ignored)
     #
     # @param callback [Block] callback that is run after the operation is done
-    def remove_deployment(hash, &callback)
+    def undeploy(hash, &callback)
```
Comment (josejulio) [inline lib/hawkular/operations/operations_api.rb:161]: I'm not sure if there is already anyone using 'remove_deployment'.
If so, we could use [this](http://stackoverflow.com/a/4763178) to create an alias and mark it as deprecated.

File: spec/integration/operations_spec.rb [hunk 2]
```diff
@@ -265,15 +265,15 @@ module Hawkular::Operations::RSpec
...
+    it 'Restart should be performed and eventually respond with success' do
       wf_server_resource_id = 'Local~~'
       alerts_war_resource_id = 'Local~%2Fdeployment%3Dhawkular-alerts-actions-email.war'
       path = CanonicalPath.new(tenant_id: @tenant_id,
                                feed_id: @feed_id,
                                resource_ids: [wf_server_resource_id, alerts_war_resource_id])
 
       redeploy = {
-        operationName: 'Redeploy',
+        operationName: 'Restart',
```
Comment (josejulio) [inline spec/integration/operations_spec.rb]: I'm just wondering. Why change Redeploy to Restart ?

File: lib/hawkular/operations/operations_api.rb [hunk 3]
```diff
@@ -128,46 +128,113 @@ def invoke_specific_operation(operation_payload, operation_name, &callback)
...
+    # Deploys an archive file into WildFly
     #
     # @param [Hash] hash Arguments for deployment
     # @option hash [String]  :resource_path canonical path of the WildFly server into which we deploy
     # @option hash [String]  :destination_file_name resulting file name
     # @option hash [String]  :binary_content binary content representing the war file
-    # @option hash [String]  :enabled whether the deployment should be enabled or not
+    # @option hash [String]  :enabled whether the deployment should be enabled immediately, or not (default = true)
+    # @option hash [String]  :force_deploy whether to replace existing content or not (default = false)
+    # @option hash [String]  :server_groups comma-separated list of server groups for the operation (default = ignored)
```
Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb:139]: Does that mean no server-group  = standalone usage? Or domain usage by only uploading into the domain-controller content repository?

**Comment by pilhuhn** [comment]
Is the backend code for this already in CmmandGW and Agent?

File: lib/hawkular/operations/operations_api.rb (hunk 1 above)
Comment (jshaughn) [inline lib/hawkular/operations/operations_api.rb:161]: I'm not sure if we're worried about back compatibility in the gem at this point.  I didn't code for it, @pilhuhn, do we need back compatibility at this time?  If so I can add it.

File: spec/integration/operations_spec.rb (hunk 2 above)
Comment (jshaughn) [inline spec/integration/operations_spec.rb]: After some meetings with product management the PRD was updated with the desired language around deployments.  The operations we'll offer are (WF terms in parens):  Deploy (add content), Undeploy (remove), Restart (redeploy), Enable (deploy) and Disable (undeploy).  I wanted to carry those terms through our code consistently from MIQ to gem to hawkular-services (command-gateway) to agent.

File: lib/hawkular/operations/operations_api.rb (hunk 1 above)
Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb:161]: On 16 Aug 2016, at 21:30, Jay Shaughnessy wrote:

> I'm not sure if we're worried about back compatibility in the gem at 
> this point.  I didn't code for it, @pilhuhn, do we need back 
> compatibility at this time?  If so I can add it.

While I do not believe anyone other than us is using the Gem,
not adding the alias is a breaking change and would need
us to release a major version of the Gem with appropriate
warnings.

## 

hwr@pilhuhn.de  m:0179/207 4919  b:http://pilhuhn.blogspot.com

File: lib/hawkular/operations/operations_api.rb (hunk 3 above)
Comment (jshaughn) [inline lib/hawkular/operations/operations_api.rb:139]: The server_groups is optional and to be supplied when we're dealing with a domain situation.  If not specified then we're talking standalone.  Basically, it is assumed that if server_groups are specified then the WF server performing the command is aware of those server_groups.  You should be able to upload content and enable it (i.e. our default for Add Deployment) whether it's standalone or applied to server groups.  Did I answer the question, I'm not sure?

**Comment by jshaughn** [comment]
> Is the backend code for this already in CmmandGW and Agent?

Yes, those PRs are already merged.   I need to check whether h services has been updated with an agent release containing the new commands.

File: lib/hawkular/operations/operations_api.rb (hunk 1 above)
Comment (jshaughn) [inline lib/hawkular/operations/operations_api.rb:161]: I'll make an alias.

File: lib/hawkular/operations/operations_api.rb (hunk 3 above)
Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb:139]: It does.

There are also two another use case (for domain mode):
- upload stuff into the domain-controller content repo. Here i guess we do not do that independently, but only via deploying to a server group
- when deploying to a server group, allowing to pick a deployment that already exists in the content repo of the DC. The use case behind is the 'first deploy to test, then promote to integration and production'.

(and with 1) also to the possibility to completely remove that content from the DC-content repo.

Comment (jshaughn) [inline lib/hawkular/operations/operations_api.rb:139]: > upload stuff into the domain-controller content repo. 

This is doable by performing 'add_deployment(enabled=false)'

> when deploying to a server group, allowing to pick a deployment that already exists in the content repo of the DC. 

This is doable if you first perform the above, 'add_deployment(enabled=false)' and later perform an 'enable'

> (and with 1) also to the possibility to completely remove that content from the DC-content repo

This is doable by performing 'undeploy('removeContent=true)'.  Note that this is the default so really just performing 'undeploy' will do this.  You can call 'undeploy('removeContent=false)', although that is really a verbose way of just performing a 'disable'.

Note that there is one more option, add_deployment(forceDeploy).  If true content is replaced if it exists, if false, the default, an error is generated.

**Comment by jshaughn** [comment]
@pilhuhn I believe this is mergeable. Now that we have the backward compatibility it should be useable with the current version of MIQ.  The new API stuff will be exercised when my MIQ PR is merged.  But that  PR is WIP until we have an updated gem and H services is released along with a new agent and inventory.

**Comment by jshaughn** [comment]
@pilhuhn please merge prior to PTO :-)

**Comment by pilhuhn** [comment]
Thanks @jshaughn !

**Pull request merged**